From 8c58320d66a78282ad9f9fd6228fa36696519d0b Mon Sep 17 00:00:00 2001
From: Phil Edworthy <phil.edworthy@renesas.com>
Date: Tue, 11 Dec 2012 11:37:06 +0000
Subject: [PATCH 105/215] SGX: Check CMA pool is entirely in one DDR bank.

Under certain situations, the SGX hardware cannot access both DDR
memory banks. Since all memory used by the SGX hardware is dma
coherent, and we use CMA to provide this, this patch adds a check
for this condition.

Signed-off-by: Phil Edworthy <phil.edworthy@renesas.com>
---
 arch/arm/mm/dma-mapping.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/arch/arm/mm/dma-mapping.c b/arch/arm/mm/dma-mapping.c
index 2b03f6f..d29e6df 100644
--- a/arch/arm/mm/dma-mapping.c
+++ b/arch/arm/mm/dma-mapping.c
@@ -267,6 +267,24 @@ static int dma_mmu_remap_num __initdata;
 
 void __init dma_contiguous_early_fixup(phys_addr_t base, unsigned long size)
 {
+#ifdef CONFIG_ARCH_R8A7779
+#define DDR_BOUNDARY 0x80000000UL
+	/*
+	 * The SGX hardware on the R-Car H1 device cannot access two banks of
+	 * DDR. Since all the memory used by the SGX is allocated by CMA, we
+	 * are checking here that the CMA region is entirely in one DDR bank.
+	 * It's not very nice to have it here, but it seems the sensible thing.
+	 */
+	if ((base < DDR_BOUNDARY) && (base+size > DDR_BOUNDARY)) {
+		pr_warn("\nr8a7779: (R-Car H1)\n"
+			"The memory used by the SGX hardware potentially "
+			"crosses the DDR boundary\nThis memory is allocated "
+			"by CMA, so either change the size of CMA, or adjust\n"
+			"the base address or size of memory available to the "
+			"kernel\n\n");
+	}
+#endif
+
 	dma_mmu_remap[dma_mmu_remap_num].base = base;
 	dma_mmu_remap[dma_mmu_remap_num].size = size;
 	dma_mmu_remap_num++;
-- 
1.7.9.5

